bin/picca_Pk1D.py

```diff
@@ -7,152 +7,367 @@
-            function of lambda_rest_frame
-        hist_weighted_delta_obs_frame: TProfile
-            Root TProfile histogram to store the mean weighted delta as a
-            function of observed lambda
-    """
+    diff_lambda = np.diff(10**delta.log_lambda)
+    diff_log_lambda = np.diff(delta.log_lambda)
+    q5_lambda, q25_lambda = np.percentile(diff_lambda, [5, 25])
+    q5_log_lambda, q25_log_lambda = np.percentile(diff_log_lambda, [5, 25])
+    if (q25_lambda - q5_lambda) < 1e-6:
+        #we can assume linear binning for this case
+        linear_binning = True
+        delta_lam = np.median(diff_lambda)
+    elif (q25_log_lambda - q5_log_lambda) < 1e-6 and q5_log_lambda < 0.01:
+        #we can assume log_linear binning for this case
+        linear_binning = False
+        delta_lam = np.median(diff_log_lambda)
+    elif (q5_log_lambda >= 0.01):
+        raise ValueError(
+            "Could not figure out if linear or log wavelength binning was used, probably submitted lambda as log_lambda"
+        )
+    else:
+        raise ValueError(
+            "Could not figure out if linear or log wavelength binning was used"
+        )
+
+    return linear_binning, delta_lam
+
+
+# loop over input files
+def process_all_files(index_file_args):
+    global num_data
+    file_index, file, args = index_file_args
+    if file_index % 5 == 0:
+        userprint("\rread {} of {} {}".format(file_index, args.len_files,
+                                              num_data),
+                  end="")
+
+    # read fits or ascii file
+    if args.in_format == 'fits':
+        hdul = fitsio.FITS(file)
+        deltas = [Delta.from_fitsio(hdu, pk1d_type=True) for hdu in hdul[1:]]
+    elif args.in_format == 'ascii':
+        ascii_file = open(file, 'r')
+        deltas = [Delta.from_ascii(line) for line in ascii_file]
+
+    #add the check for linear binning on first spectrum only (assuming homogeneity within the file)
+    delta = deltas[0]
+    linear_binning, delta_lam = check_linear_binning(delta)
+    if linear_binning:
+        userprint("\n\nUsing linear binning, results will have units of AA")
+        delta_lambda = delta_lam
+        if (args.disable_reso_matrix or not hasattr(delta, 'resolution_matrix')
+                or delta.resolution_matrix is None):
+            userprint(
+                "Resolution matrix not found or disabled, using Gaussian resolution correction\n"
+            )
+            reso_correction = "Gaussian"
+        else:
+            userprint("Using Resolution matrix for resolution correction\n")
+            reso_correction = "matrix"
+    else:
+        userprint("\n\nUsing log binning, results will have units of km/s")
+        delta_log_lambda = delta_lam
+        reso_correction = "Gaussian"
+        userprint("Using Gaussian resolution correction\n")
+
+    num_data += len(deltas)
+    userprint("\n ndata =  ", num_data)
+    results = None
+
+    def process_file(delta):
```

Comment:
originally I wanted to parallelize over the different deltas in a file, but that requires moving more things around (as the function needs to be definable outside main), atm I'm parallelizing over the whole thing to avoid the complete overhaul (it's not that important yet to parallelize except if running on yr5 predictions) so the function could actually become a for loop again.